dynamoprojection/options.go

```diff
@@ -0,0 +1,156 @@
+// handler.
+type HandlerOption interface {
+	applyOptionToAdaptor(*decorators)
+}
+
+// ResourceRepositoryOption is used to alter the behavior ResourceRepository.
+type ResourceRepositoryOption interface {
+	applyResourceRepositoryOption(*decorators)
+}
+
+// TableOption is used to alter the behavior operations related to table
+// manipulations.
+type TableOption interface {
+	applyTableOption(*decorators)
+}
+
+type decorators struct {
+	decorateGetItem            func(*dynamodb.GetItemInput) []func(*dynamodb.Options)
+	decoratePutItem            func(*dynamodb.PutItemInput) []func(*dynamodb.Options)
+	decorateDeleteItem         func(*dynamodb.DeleteItemInput) []func(*dynamodb.Options)
+	decorateTransactWriteItems func(*dynamodb.TransactWriteItemsInput) []func(*dynamodb.Options)
+	decorateCreateTableItem    func(*dynamodb.CreateTableInput) []func(*dynamodb.Options)
+	decorateDeleteTableItem    func(*dynamodb.DeleteTableInput) []func(*dynamodb.Options)
+}
+
+type options struct {
+	applyOptionToAdaptorFunc          func(*decorators)
+	applyResourceRepositoryOptionFunc func(*decorators)
+	applyTableOptionFunc              func(*decorators)
+}
+
+func (o *options) applyOptionToAdaptor(decorators *decorators) {
+	if o.applyOptionToAdaptorFunc != nil {
+		o.applyOptionToAdaptorFunc(decorators)
+	}
+}
+
+func (o *options) applyResourceRepositoryOption(decorators *decorators) {
+	if o.applyResourceRepositoryOptionFunc != nil {
+		o.applyResourceRepositoryOptionFunc(decorators)
+	}
+}
+
+func (o *options) applyTableOption(decorators *decorators) {
+	if o.applyTableOptionFunc != nil {
+		o.applyTableOptionFunc(decorators)
+	}
+}
+
+// WithDecorateGetItem adds a decorator for GetItem operation. The decorator can
+// modify the passed GetItemInput structure and return a slice of request.Option
+// to alter the request prior to its execution.
+func WithDecorateGetItem(
+	dec func(*dynamodb.GetItemInput) []func(*dynamodb.Options),
+) interface {
+	HandlerOption
+	ResourceRepositoryOption
+} {
+	return &options{
+		applyOptionToAdaptorFunc: func(d *decorators) {},
+		applyResourceRepositoryOptionFunc: func(d *decorators) {
+			d.decorateGetItem = dec
+		},
+	}
+}
+
+// WithDecoratePutItem adds a decorator for PutItem operation. The decorator can
+// modify the passed PutItemInput structure and return a slice of request.Option
+// to alter the request prior to its execution.
+func WithDecoratePutItem(
+	dec func(*dynamodb.PutItemInput) []func(*dynamodb.Options),
+) interface {
+	HandlerOption
+	ResourceRepositoryOption
+} {
+	return &options{
+		applyOptionToAdaptorFunc: func(d *decorators) {},
+		applyResourceRepositoryOptionFunc: func(d *decorators) {
+			d.decoratePutItem = dec
+		},
+	}
+}
+
+// WithDecorateDeleteItem adds a decorator for DeleteItem operation. The
+// decorator can modify the passed DeleteItemInput structure and return a slice
+// of request.Option to alter the request prior to its execution.
```

Comment:
Please see suggested format on `WithDecorateGetItem`.

Reply:
Should be fixed in [5b85c8c](https://github.com/dogmatiq/projectionkit/pull/135/commits/5b85c8cd23a15394ccf4a704e5f4b3b26a2800fc)